### DISCUSSION TITLE
Reading request body in AuthenticationFetcher implementations with Micronaut 2.5.x

### DISCUSSION THREAD

**Discussion by adamkobor** [micronaut-projects/micronaut-security#649] [Q&A]
Hey there, I have a question about accessing the request body. 

We have to integrate with a 3rd party service's webhooks that uses a quite unusual signing method for its requests. Long story short: we have to read the request's body as a string to be able to check the signature they provide in a dedicated header. Before Micronaut 2.5.x we used a custom AuthenticationFetcher implementation, something like this:

```kotlin
@Singleton
class CustomAuthFetcher(private val config: CustomAuthConfig) : AuthenticationFetcher {

    override fun fetchAuthentication(request: HttpRequest<*>): Flowable<Authentication> {
        val bodySignature = request.body.map { body ->
            // Do something with the body
        }.orElse("")

       return if (bodySignature == request.headers.get("X-SIGNATURE") {
           Authentication(name = "CustomName", roles = setOf("ROLE_CUSTOM")) 
       } else Flowable.empty()
    }
}
```
This solution worked without any problem, and we could use the `@Secured` annotation with the custom role in our controllers. 
However, in Micronaut 2.5.x it won't work, because `request.body` is always an empty `Optional`. I understand that this is the desired behavior, but I'm also curious if there is an idiomatic way to achieve the original behavior?
After the upgrade we've refactored our logic, and we are just bypassing the built-in security checks by providing `@Secured("isAnonymous()")` to the controller methods and checking the body inside these methods:

```kotlin
@Secured("isAnonymous()")
override fun event(
    @Body request: EventRequest,
    rawRequest: HttpRequest<String>
) {
    customAuthenticator.authenticate(rawRequest) // This one throws an exception in case of an authorization error and we transform it into a 401
    // ... 
}
```

While this approach also works, it seems a bit strange and "hackish" :(

**Comment by sdelamo** [comment]
@jameskleeh can you comment in this issue?

**Comment by tcrespog** [comment]
Likely related issue: https://github.com/micronaut-projects/micronaut-security/issues/689.

I opened that issue because the body was present at that stage in the 2.4.x line and I couldn't find any justification for the change of behavior in the docs; therefore, I considered it was a bug. If it's not, an explanation would be appreciated.

**Comment by jameskleeh** [comment]
This could be done via a constraint or AOP. A constraint makes more sense to me from a logical perspective, but if you want a custom response AOP might make that easier since you can throw a custom exception.

**Comment by garryturk** [comment]
This bit me today upgrading a project from 2.4.x to 2.5.x. My scenario is an app receiving requests from machine clients so each request is individually authorised by a signature in a header.
To calculate that signature requires fields from the body so we have a few SecurityRules in sequence that currently implement this. It isn't clear to me from the above discussions what the idiomatic way to handle this scenario would be - looks like SecurityRules post 2.5.x are completely off the table now.
Obviously I'd rather do this validation before my controller methods are invoked so would really rather not have to rework those to explicitly include this validation logic. But in the absence of authenticated users I'm not sure what the best approach for per-request signature body validation looks like?

Thanks!

**Reply by tcrespog** [reply]
I experienced the same issue with a security rule (that's why I opened https://github.com/micronaut-projects/micronaut-security/issues/689), and following https://github.com/micronaut-projects/micronaut-security/discussions/649#discussioncomment-966274 I transformed the security rule into an AOP interceptor.

These are the relevant pieces of code to implement an AOP interceptor that replaces a security rule — applicable through an annotation (just like the built-in [`SecuredAnnotationRule`](https://github.com/micronaut-projects/micronaut-security/blob/master/security/src/main/java/io/micronaut/security/rules/SecuredAnnotationRule.java)) — and grants access when there's a body present in the request:
<details>
<summary>Annotation</summary>

```java
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Around
@interface MySecured {
}
```
</details>

<details>
<summary>Security rule (now disabled in favor of the interceptor)</summary>

```java
@Singleton
public class MySecuredAnnotationRule implements SecurityRule {

    /**
     * It has priority over the built-ins
     * {@link io.micronaut.security.rules.SecuredAnnotationRule} and {@link io.micronaut.security.rules.ConfigurationInterceptUrlMapRule} in case several apply.
     */
    public static final int ORDER = Math.min(SecuredAnnotationRule.ORDER, ConfigurationInterceptUrlMapRule.ORDER) - 1;

    @Override
    public int getOrder() {
        return ORDER;
    }

    @Override
    public SecurityRuleResult check(HttpRequest<?> request, @Nullable RouteMatch<?> routeMatch, @Nullable Map<String, Object> claims) {
        
        // check whether the rule applies
        if ( !(routeMatch instanceof UriRouteMatch) ) {
            return SecurityRuleResult.UNKNOWN;
        }

        final AnnotationValue<MySecured> annot = routeMatch.getAnnotation(MySecured);
        if (!annot) {
            return SecurityRuleResult.UNKNOWN;
        }

        // check if the body is present (never present since MN 2.5)
        Optional<?> body = request.getBody();
        return body.isPresent() ? SecurityRuleResult.ALLOWED : SecurityRuleResult.REJECTED;
    }
}
```
</details>

<details>
<summary>AOP Interceptor</summary>

```java
@Singleton
@InterceptorBean(MySecured)
class MySecuredInterceptor implements MethodInterceptor<Object, Object> {

    /**
     * Here you can control the order of execution in regards to the built-in interceptors,
     * for example here I'm interested in executing before the one associated with @Transactional 
     */
    public static final int ORDER = InterceptPhase.TRANSACTION.position - 1;

    @Override
    public int getOrder() {
        return ORDER;
    }

    @Override
    public Object intercept(MethodInvocationContext<Object, Object> context) {

        // you can get the annotation and retrieve member values from it if needed as in the security rule
        final AnnotationValue<MySecured> annot = context.<MySecured>getAnnotation(MySecured);

        // this way you can access the current request
        Optional<HttpRequest<?>> request = ServerRequestContext.currentRequest().orElse(null);
        if (request == null) 
            throw new IllegalStateException("Request not found. The interceptor works on controller methods only");

        // check if the body is present
        Optional<?> body = request.getBody();
        if (!request.body.isPresent())
            // the rejection flow is controlled through an exception. I map this custom exception to a 403 response in a global error handler (@Error)
            throw new ForbiddenException();

        // allowed
        return context.proceed();
    }
}
```
</details>

<details>
<summary>Usage</summary>

```java
@Controller("/pet")
class PetController {

    @Post("/")
    @MySecured
    public HttpResponse post(Pet pet) {
        return HttpResponse.ok();
    }

}
```
</details>

I was able to achieve the same outcome I had with the security rule. The only restriction I can see is that the interceptor will always execute after the security rules (whereas you can specify your custom security rule to run before the others), but I suppose you may work around that.

Hope it helps.

**Reply by adamkobor** [reply]
Thanks @tcrespog for the detailed explanation! <3

**Comment by jameskleeh** [comment]
Garry - I’ll put together an example for you

**Comment by garryturk** [comment]
Thanks James - any pointers would be gratefully received!

**Comment by sdelamo** [comment]
I've created a PR with an example: https://github.com/micronaut-projects/micronaut-security/pull/756

It is similar to @tcrespog suggestion.

**Comment by garryturk** [comment]
Thanks to @tcrespog  and @sdelamo   for the pointers. Transformed my rule into an interceptor and it works fine.

Much appreciated, hugely useful!
